redshift_sqlalchemy/dialect.py

```diff
@@ -487,3 +487,105 @@ def visit_bindparam(bindparam, compiler, **kw):
+        for child in root.get_children():
+            yc = gen_columns_from_children(child)
+            for it in yc:
+                yield it
+    elif isinstance(root, sa.Column):
+        yield root
+
+
+@compiles(Delete, 'redshift')
+def visit_delete_stmt(element, compiler, **kwargs):
+    """
+    Adds redshift-dialect specific compilation rule for the
+    delete statement.
+
+    Redshift DELETE syntax can be found here:
+    http://docs.aws.amazon.com/redshift/latest/dg/r_DELETE.html
+
+    .. :code-block: sql
+
+        DELETE [ FROM ] table_name
+        [ { USING } table_name, ...]
+        [ WHERE condition ]
+
+    By default, SqlAlchemy compiles DELETE statements with the
+    syntax:
+
+    .. :code-block: sql
+
+        DELETE [ FROM ] table_name
+        [ WHERE condition ]
+
+    problem illustration:
+
+    >>> from sqlalchemy import Table, Column, Integer, MetaData, delete
+    >>> from redshift_sqlalchemy.dialect import RedshiftDialect
+    >>> meta = MetaData()
+    >>> table1 = Table(
+    ... 'table_1',
+    ... meta,
+    ... Column('pk', Integer, primary_key=True)
+    ... )
+    ...
+    >>> table2 = Table(
+    ... 'table_2',
+    ... meta,
+    ... Column('pk', Integer, primary_key=True)
+    ... )
+    ...
+    >>> del_stmt = delete(table1).where(table1.c.pk==table2.c.pk)
+    >>> str(del_stmt.compile(dialect=RedshiftDialect()))
+    'DELETE FROM table_1 USING table_2 WHERE table_1.pk = table_2.pk'
+    >>> str(del_stmt)
+    'DELETE FROM table_1 WHERE table_1.pk = table_2.pk'
+    >>> del_stmt2 = delete(table1)
+    >>> str(del_stmt2)
+    'DELETE FROM table_1'
+    >>> del_stmt3 = delete(table1).where(table1.c.pk > 1000)
+    >>> str(del_stmt3)
+    'DELETE FROM table_1 WHERE table_1.pk > :pk_1'
+    >>> str(del_stmt3.compile(dialect=RedshiftDialect()))
+    'DELETE FROM table_1 WHERE table_1.pk >  %(pk_1)s'
+    """
+
+    # Set empty strings for the default where clause and using clause
+
+    whereclause = ''
+    usingclause = ''
+
+    # determine if the delete query needs a ``USING`` injected
+    # by inspecting the whereclause's children & their children...
+    # first, the where clause text is buit, if applicable
+    # then, the using clause text is built, if applicable
+    # note:
+    #   the tables in the using clause are sorted in alphabetical order
+    #   simply to guarantee that unit tests don't fail uncessarily
+
+    whereclause_tuple = element.get_children()
+    if whereclause_tuple:
+        usingclause_set = set()
+        whereclause = ' WHERE {clause}'.format(clause=compiler.process(*whereclause_tuple))
+
+        whereclause_columns = gen_columns_from_children(element)
+        for col in whereclause_columns:
+            usingclause_set.add(_tablename(col.table, compiler))
+        usingclause_set.remove(_tablename(element.table, compiler))
+        if usingclause_set:
+            usingclause = ' USING {clause}'.format(clause=','.join(sorted(usingclause_set)))
```

Comment:
rather than using a sort I'd probably go for the order the elements appear in the CTE:

eg 
```python
usingclause = []
usingclause_set = set():
for col in whereclause_columns:
    table = _tablename(col.table, compiler)
    if table not in usingclause_set:
       usingcluase.append(table)
       usingclause.add(table)